### PR TITLE
Add newdocs functionality

### LINKED ISSUE #204: Show Latest Documents
Would be nice to show which documents were recently added or modified on the front page

### PR DESCRIPTION
**Pull request by lauterb**
* fixes #204

### CODE REVIEW

File: CorpusSearch/Services/OpenDataLoader.cs [hunk 1]
```diff
@@ -60,6 +62,34 @@ public static List<string> GetJsonPaths(LoadConfig lConfig)
...
+        {
+            var latestDocumentsFile = Path.Combine(
+                AppDomain.CurrentDomain.BaseDirectory,
+                "OpenData",
+                "newdocs.txt"
+            );
+             
+            var lines = await File.ReadAllLinesAsync(latestDocumentsFile);
+
+            var dateAndPath = lines.Select(x =>
```
Comment (david-allison) [inline CorpusSearch/Services/OpenDataLoader.cs:78]: `x` -> `line`

File: CorpusSearch/Services/OpenDataLoader.cs [hunk 2]
```diff
@@ -60,6 +62,34 @@ public static List<string> GetJsonPaths(LoadConfig lConfig)
...
+                (
+                    Date: DateTime.Parse(x[.."2024-02-27 04:50:38".Length]),
+                    Path: x["2024-02-27 04:50:38 ".Length..]
```
Comment (david-allison) [inline CorpusSearch/Services/OpenDataLoader.cs:81]: I feel this is fairly readable, but it could do with a comment with a sample line

File: CorpusSearch/Services/OpenDataLoader.cs [hunk 3]
```diff
@@ -60,6 +62,34 @@ public static List<string> GetJsonPaths(LoadConfig lConfig)
...
+
+        
+
+        public static async Task<List<(IDocument, DateTime)>> LoadLatestDocuments(WorkService workService)
+        {
+            var latestDocumentsFile = Path.Combine(
+                AppDomain.CurrentDomain.BaseDirectory,
+                "OpenData",
+                "newdocs.txt"
+            );
```
Comment (david-allison) [inline CorpusSearch/Services/OpenDataLoader.cs:74]: comment an expected path above for brevity

Make a note on `newdocs.txt` is generated

File: CorpusSearch/Services/OpenDataLoader.cs [hunk 4]
```diff
@@ -60,6 +62,34 @@ public static List<string> GetJsonPaths(LoadConfig lConfig)
...
+             
+            var lines = await File.ReadAllLinesAsync(latestDocumentsFile);
```
Comment (david-allison) [inline CorpusSearch/Services/OpenDataLoader.cs:76]: What happens on an exception here?

What should happen?

File: CorpusSearch/Services/OpenDataLoader.cs [hunk 5]
```diff
@@ -60,6 +62,34 @@ public static List<string> GetJsonPaths(LoadConfig lConfig)
...
+            var dateAndPath = lines.Select(x =>
+                (
+                    Date: DateTime.Parse(x[.."2024-02-27 04:50:38".Length]),
+                    Path: x["2024-02-27 04:50:38 ".Length..]
+                )
+            );
+            
+            var allDocuments = await workService.GetAll();
+
+            var map = allDocuments.ToDictionary(x => x.RelativeCsvPath, x => x);
```
Comment (david-allison) [inline CorpusSearch/Services/OpenDataLoader.cs:87]: single comment somewhere explaining that we're looking up the filepaths defined in `newdocs.txt` to obtain the metadata for the file at the path

Comment (david-allison) [inline CorpusSearch/Services/OpenDataLoader.cs:87]: rename `map` to make the intention more legible

File: CorpusSearch/Services/OpenDataLoader.cs [hunk 6]
```diff
@@ -134,11 +164,11 @@ private string RelativeLocationOnDisk
         }
 
         public override string GitHubRepo { get; set; }
-        public override string RelativeCsvPath => RelativeLocationOnDisk + "\\" + CsvFileName;
+        public override string RelativeCsvPath => RelativeLocationOnDisk + Path.DirectorySeparatorChar + CsvFileName;
 
         public override string ToString()
         {
-            return RelativeLocationOnDisk + "\\manifest.json";
+            return RelativeLocationOnDisk + Path.DirectorySeparatorChar + "manifest.json";
```
Comment (david-allison) [inline CorpusSearch/Services/OpenDataLoader.cs:171]: Could be split into a separate commit beforehand

File: CorpusSearch/Startup.cs [hunk 7]
```diff
@@ -95,14 +101,21 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env, WorkServ
             {
                 Console.WriteLine("Failed to init anonymous analytics. Was CORPUS_SEARCH_SEGMENT_KEY set?");
             }
+
+            Console.WriteLine("Envs:");
+            foreach (var c in Configuration.AsEnumerable())
+            {
+                Console.WriteLine(c.Key + " = " + c.Value);
+            }
```
Comment (david-allison) [inline CorpusSearch/Startup.cs:109]: ⚠️ Printing environment variables to the log is how servers get owned

File: CorpusSearch/newdocs [hunk 8]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+#note: this script needs to be run on the manx-search-data repo
+# not the repo it lives in
```
Comment (david-allison) [inline CorpusSearch/newdocs:3]: because it needs to access the git history of `manx-search-data`

Comment (david-allison) [reply CorpusSearch/newdocs:3]: It would be useful to have this script stop execution if this condition is not met

File: CorpusSearch/newdocs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+# weirdest trick is the pattern matching to get dir from fdir
+# basically ## means delete the longest matching pattern from the var, # the shortest
+# for more on that see https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
```
Comment (david-allison) [inline CorpusSearch/newdocs:6]: Move this to where the functionality is executed

Comment (david-allison) [reply CorpusSearch/newdocs:6]: Split up the 'what' and the 'how'. Most users are reading to skim, so provide this information upfront

* Explain the intention of the code
* Give an example
* Explain how it works

File: CorpusSearch/newdocs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+IGNORE_RENAME=--diff-filter=r
+#substitute % for space because the latter is used as an array delimiter by the shell
+for  f in $(git log --since=${SINCE} ${IGNORE_RENAME} --name-only | grep ^OpenData | grep document.csv$ |  sort -u | tr ' ' '%')
+do
+    ff=$(echo $f |  tr '%' ' ')
+
+    if [[ -z $(git ls-files "$ff") ]]; then continue; fi
+
+    cd=$(git log -1 --date=format-local:'%Y-%m-%d %H:%M:%S' --pretty=format:"%cd" "$ff")
```
Comment (david-allison) [inline CorpusSearch/newdocs:46]: nit: list the output

File: CorpusSearch/newdocs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+    if [[ 0 -eq $DIR_ONLY ]]; then
+        ffs=$(echo $cd $ff)
+    else
+        fdir=$(echo $ff | sed -e 's/\/document.csv$//')
+        dir=${fdir##*/} #deletes longest matching pattern
+        ffs=$(echo $cd $dir)
+    fi
+
+    echo $ffs
```
Comment (david-allison) [inline CorpusSearch/newdocs:56]: nit: list the output

File: CorpusSearch/newdocs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+# basically ## means delete the longest matching pattern from the var, # the shortest
+# for more on that see https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+
+MAX=20
+SINCE=three-months-ago
+DIR_ONLY=0
+usage="$(basename "$0") [-d] [-h] [-m MAX] [-s SINCE]  - list recently changed docs when run in manx-search-data
+
+options are:
+    -d  print the last directory only
```
Comment (david-allison) [inline CorpusSearch/newdocs:14]: I don't understand what this does, or the reason it's necessary. Can it be removed?

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    const newarray: JSX.Element[] = []
+    for ( let i=0; i < len; i++) {
+        const {name,ident} = newDocs[i]
+        const link = <CorpusLink key={name} name={name} ident={ident} />
+        newarray.push(link)
+    }
+    return (<div className="Newbies">
+        <header className="Newbies-header">
+        </header>
+        <div>New and Featured Transcriptions:<br/>
```
Comment (david-allison) [inline CorpusSearch/ClientApp/src/routes/Newbies.tsx:40]:
```suggestion
        <div>Recent Additions:<br/>
```

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+            <div>{newarray}</div>
```
Comment (david-allison) [inline CorpusSearch/ClientApp/src/routes/Newbies.tsx:41]: I'd consider an unordered list:` <ul><li>item1</li><li>item2</li></ul>`

File: CorpusSearch/ClientApp/src/routes/Home.tsx [hunk 15]
```diff
@@ -150,20 +150,26 @@ export const Home = () => {
             {hasNoSearch && stats && <span className={"homeText"}>
                 {stats != "error" ?
                     <span className={"homeText"} style={{textAlign: "center"}}>
-                        <span style={{display: "inline"}}>Search our growing collection of over <b title={`${stats.uniqueManxWordCount.toLocaleString()} unique words`}>{stats.manxWordCount.toLocaleString()} Manx words</b> or&nbsp;<a href={"/Browse"}>browse&nbsp;{stats.documentCount.toLocaleString()} documents</a></span>
-                    </span>    
-                :
+                        <span style={{display: "inline"}}>Search our growing collection of over <b
+                            title={`${stats.uniqueManxWordCount.toLocaleString()} unique words`}>{stats.manxWordCount.toLocaleString()} Manx words</b> or&nbsp;
```
Comment (david-allison) [inline CorpusSearch/ClientApp/src/routes/Home.tsx:154]: revert the auto-formatting changes.

If they're necessary, they should go in their own commit

**Review by david-allison — CHANGES_REQUESTED** [review]
Thanks!

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    const fetchNewDocs = async () => {
+        await fetch("api/metadata/latest")
+            .then((res)=> {
+                if (!res.ok) {
+                    console.warn(`metadata/latest returned status: ${res.status}`)
+                }
+                return res.json()
+            }).then((json: DocType[])=> {
+                setNewDocs(json)
+        })
```
[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:24]: Sure, but would have to get the casting to DocType[] correct.

[the remark this answers is not in the document]
Comment (david-allison) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:24]: You should be able to use `as DocType[]`

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 13 above)
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:40]: Fine. This was a placeholder name anyway. Only it's not just additions but also changes

File: CorpusSearch/ClientApp/src/routes/Home.tsx (hunk 15 above)
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Home.tsx:154]: Okay

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 14 above)
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:41]: What does this get us?

File: CorpusSearch/newdocs (hunk 12 above)
Comment (lauterb) [reply CorpusSearch/newdocs:14]: It's much easier for me to read that output to see what changed than read the whole path

File: CorpusSearch/newdocs (hunk 11 above)
Comment (lauterb) [reply CorpusSearch/newdocs:56]: In the usage?

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:41]: Okay

File: CorpusSearch/newdocs (hunk 11 above)
Comment (david-allison) [reply CorpusSearch/newdocs:56]: As a comment

File: CorpusSearch/newdocs (hunk 12 above)
Comment (david-allison) [reply CorpusSearch/newdocs:14]: `-d`

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 13 above)
Comment (david-allison) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:40]: It should ideally be additions, otherwise an automated edit would mean that all files are listed

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 14 above)
Comment (david-allison) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:41]: Bullet points are better-spaced by default

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:24]: Will try. Otherwise TS2345: Argument of type 'Promise<any>' is not assignable to parameter of type 'SetStateAction<DocType[]>'.

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    }
+    fetchNewDocs().catch((e:Error)=> console.warn(`fetchNewDocs returned error:" ${e.message}`))
+},[]
```
[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:27]: What autoformatter should I be using?

File: CorpusSearch/newdocs (hunk 12 above)
Comment (lauterb) [reply CorpusSearch/newdocs:14]: Again, just so much easier to read that output

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 14 above)
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:41]: Also how to combine that with your other comment about the map?

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:24]: Doesn't that too much either:
<html>TS2352: Conversion of type 'Promise&lt;any&gt;' to type 'DocType[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.<br/>Type 'Promise&lt;any&gt;' is missing the following properties from type 'DocType[]': length, pop, push, concat, and 29 more.

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 14 above)
Comment (david-allison) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx:41]: Something like:

```tsx
<ul>
    {newDocs.map(doc => <li key={doc.name}><CorpusLink name={doc.name} ident={doc.ident} /></li>)}
</ul>
```

Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx]: Okay, just pushed a bunch of other changes, can try this next I guess

[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx]: Done

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import React, {useEffect, useState} from "react"
+function CorpusLink({ident="",name=""}): JSX.Element {
+    if ( name.slice(0,2) != "🎥"){
+        name = "📖" + name
+    }
+    return <div>
+        <a href={"docs/" + ident}>{name}</a>
+    </div>
+}
+export const Newbies = () => {
```
[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx]: Done

File: CorpusSearch/newdocs (hunk 11 above)
Comment (lauterb) [reply CorpusSearch/newdocs]: Done

File: CorpusSearch/newdocs (hunk 12 above)
Comment (lauterb) [reply CorpusSearch/newdocs]: Done

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 13 above)
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx]: Done although I changed the wording slightly

File: CorpusSearch/ClientApp/src/routes/Home.tsx (hunk 15 above)
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Home.tsx]: Done

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx]: I don't know the "best" way to format this (yet), but I just manually formatted it in a way I could live with.

[the remark this answers is not in the document]
Comment (lauterb) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx]: Also was able to test manually by doing these separately and looking at the warnings on the console:
* changing the url to break so that res.ok was false
* throwing an error and catching it

**Comment by lauterb** [comment]
Thinking "Search our growing..."  should be left-justified to match everything below it

File: CorpusSearch/ClientApp/src/routes/NewDocList.tsx [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+ */
+import React, {useEffect, useState} from "react"
+function CorpusLink({ident="",name=""}): JSX.Element {
+    if ( name.slice(0,2) != "🎥"){
+        name = "📖" + name
+    }
+    return <div>
+        <a href={"docs/" + ident}>{name}</a>
+    </div>
+}
```
Comment (david-allison) [inline CorpusSearch/ClientApp/src/routes/NewDocList.tsx]: This looks better visually if the emoji are not underlined
```patch
Index: CorpusSearch/ClientApp/src/routes/NewDocList.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/CorpusSearch/ClientApp/src/routes/NewDocList.tsx b/CorpusSearch/ClientApp/src/routes/NewDocList.tsx
--- a/CorpusSearch/ClientApp/src/routes/NewDocList.tsx	(revision 34a9dabe8c8650db9d3a8f334c6bf50af88d6ea3)
+++ b/CorpusSearch/ClientApp/src/routes/NewDocList.tsx	(date 1715022009532)
@@ -3,14 +3,25 @@
  * Which means those document.csv files that git detects as having recent changes
  */
 import React, {useEffect, useState} from "react"
-function CorpusLink({ident="",name=""}): JSX.Element {
-    if ( name.slice(0,2) != "🎥"){
-        name = "📖" + name
+
+/**
+ * Produces a link to a document, with an emoji denoting the type of content
+ * @example "🎥 Jack as Ned" as a link; no link underline for "🎥 "
+ */
+function CorpusLink({ident = "", name = ""}): JSX.Element {
+    let emojiPrefix = "📖"
+    if (name.slice(0, 2) == "🎥") {
+        name = name.substring(2).trim() // trim the emoji from the name (and any leading spaces)
+        emojiPrefix = "🎥"
     }
+    const href = "docs/" + ident
     return <div>
-        <a href={"docs/" + ident}>{name}</a>
+        {/* Better visuals if the prefix is not underlined; maintain the link for UX */}
+        <a style={{textDecoration: "none"}} href={href}>{emojiPrefix}&nbsp;</a>
+        <a href={href}>{name}</a>
     </div>
 }
+
 export const NewDocList = () => {
 type DocType =  {name:string, ident:string, uploaded:string}
 const [newDocs, setNewDocs  ] 
```

Screenshot 2024-05-06 at 19 59 42

File: CorpusSearch/ClientApp/src/routes/NewDocList.tsx [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+export const NewDocList = () => {
+type DocType =  {name:string, ident:string, uploaded:string}
```
Comment (david-allison) [inline CorpusSearch/ClientApp/src/routes/NewDocList.tsx]: TODO/Extension: we can use 'uploaded' to provide a relative date

File: CorpusSearch/ClientApp/src/routes/NewDocList.tsx [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    fetchNewDocs().catch((e:Error)=> console.warn(`fetchNewDocs returned:" ${e.message}`))
+},[]
+    )
+    const len = newDocs.length
+    if (!len) return <></>
+    return (<div className="NewDocList">
+        <header className="NewDocList-header">
+        </header>
+        <div>Recent Additions and Updates:<br/>
```
Comment (david-allison) [inline CorpusSearch/ClientApp/src/routes/NewDocList.tsx]: Don't make people read

Not sure about the bolding - your call

```suggestion
        <div><b>Recent changes:</b><br/>
```

File: CorpusSearch/ClientApp/src/routes/Home.tsx [hunk 22]
```diff
@@ -159,6 +159,7 @@ export const Home = () => {
                         </span>
                     </>
                 }
+                <div><NewDocList/></div>
                 <span style={{display: "inline", marginTop: "2em"}}>Support our revitalisation efforts by <a href={"/MailingList"}>signing up for our mailing list</a>. We'll email once in a while with updates to the corpus & other projects.</span>
```
Comment (david-allison) [inline CorpusSearch/ClientApp/src/routes/Home.tsx:163]:
```suggestion
                <span style={{display: "inline", marginTop: "1em"}}>Support our revitalisation efforts by <a href={"/MailingList"}>signing up for our mailing list</a>. We'll email once in a while with updates to the corpus & other projects.</span>
```

File: CorpusSearch/newdocs [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+for  f in $(git log --since=${SINCE} ${IGNORE_RENAME} --name-only | grep ^OpenData | grep document.csv$ |  sort -u | tr ' ' '%')
+do
+    ff=$(echo $f |  tr '%' ' ')
+
+    if [[ -z $(git ls-files "$ff") ]]; then continue; fi
+
+    cd=$(git log -1 --date=format-local:'%Y-%m-%d %H:%M:%S' --pretty=format:"%cd" "$ff")
+
+    echo $cd $ff
+    #outputs local time and the full path of each document.csv
```
Comment (david-allison) [inline CorpusSearch/newdocs:45]:
```suggestion
    # 2024-02-27 04:50:38 OpenData/miscellaneous religious printed 1600 - 1900/Wilsons Sermons/document.csv
    echo $cd $ff
```

File: CorpusSearch/Startup.cs (hunk 7 above)
Comment (david-allison) [reply CorpusSearch/Startup.cs:109]: This is still an issue

**Review by david-allison — CHANGES_REQUESTED** [review]
Looks good! final round of testing after these

File: CorpusSearch/ClientApp/src/routes/Newbies.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (david-allison) [reply CorpusSearch/ClientApp/src/routes/Newbies.tsx]: As noted in another comment, Rider's is /mostly/ decent, there was a couple of manual fixups required

The main issue here is indentation, which any auto-formatter should get close to perfect

**Review by david-allison — APPROVED** [review]
Force pushed the final round, looks good!
